File: 525.contiguous-array.cpp
/*
 * @lc app=leetcode id=525 lang=cpp
 *
 * [525] Contiguous Array
 */
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;
// @lc code=start
class Solution
{
public:
    int findMaxLength(vector<int> &nums)
    {
        // key: sum_val value: index
        unordered_map<int, int> umap;
        umap[0] = -1;
        int res = 0;
        int sum_val = 0;
        int index = 0;
        for (auto n : nums)
        {
            if (n == 0)
            {
                --sum_val;
            }
            else
            {
                ++sum_val;
            }
            if (umap.count(sum_val))
            {
                res = max(res, index - umap[sum_val]);
            }
            else
            {
                umap[sum_val] = index;
            }
            ++index;
        }
        return res;
    }
};
// @lc code=end

File: py/525.contiguous-array.py
#
# @lc app=leetcode id=525 lang=python3
#
# [525] Contiguous Array
#

# @lc code=start
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        # key: sum_val value: index
        umap = {}
        res = 0
        umap[0] = -1
        sum_val = 0
        index = 0
        for n in nums:
            if n == 0:
                sum_val -= 1
            else:
                sum_val += 1
            if sum_val in umap:
                res = max(res, index-umap[sum_val])
            else:
                u = umap[sum_val] = index
            index += 1
        return res
# @lc code=end
